File: examples/java/validation_custom.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/**
 * @file validation_custom.java
 *
 * This file contains an example demonstrating how to implement additional
 * Message validation logic in addition to that which the GMSEC API provides.
 */

import gov.nasa.gsfc.gmsec.api.*;

import gov.nasa.gsfc.gmsec.api.field.Field;
import gov.nasa.gsfc.gmsec.api.field.StringField;

import gov.nasa.gsfc.gmsec.api.mist.gmsecMIST;
import gov.nasa.gsfc.gmsec.api.mist.ConnectionManager;
import gov.nasa.gsfc.gmsec.api.mist.MessageValidator;

import gov.nasa.gsfc.gmsec.api.mist.message.MistMessage;

import gov.nasa.gsfc.gmsec.api.util.Log;

import java.util.ArrayList;


class CustomMessageValidator extends MessageValidator
{
	public Status validateMessage(Message msg)
	{
		Status status = new Status();

		try
		{
			//o Get message type and subtype
			String type    = "";
			String subtype = "";

			if (msg.hasField("MESSAGE-TYPE"))
			{
				type = msg.getStringValue("MESSAGE-TYPE");
			}
			if (msg.hasField("C2CX-SUBTYPE"))
			{
				subtype = msg.getStringValue("C2CX-SUBTYPE");
			}
			else if (msg.hasField("MESSAGE-SUBTYPE"))
			{
				subtype = msg.getStringValue("MESSAGE-SUBTYPE");
			}

			//o Ensure we have a Heartbeat message and it contains the PUB-RATE field
			if (type.equals("MSG") && subtype.equals("HB") && msg.hasField("PUB-RATE"))
			{
				long pubRate = msg.getIntegerValue("PUB-RATE");

				//o Ensure PUB-RATE field value is between 10 and 60 (inclusive)
				if (pubRate < 10 || pubRate > 60)
				{
					status.set(StatusClassification.MSG_ERROR, StatusCode.VALUE_OUT_OF_RANGE, "PUB-RATE field does not have a valid value");
				}
			}
			else
			{
				status.set(StatusClassification.MSG_ERROR, StatusCode.INVALID_MSG, "Non-Heartbeat message received");
			}
		}
		catch (GMSEC_Exception e)
		{
			status = new Status(e.getErrorClassification(), e.getErrorCode(), e.getErrorMessage(), e.getCustomCode());
		}

		return status;
	}
}


public class validation_custom
{
	private static final String HB_MSG_SUBJECT     = "GMSEC.MISSION.SATELLITE.MSG.HB.VALIDATION-CUSTOM";
	private static final String ALT_HB_MSG_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.C2CX.VALIDATION-CUSTOM.HB";


	public static void main(String[] args)
	{
		if (args.length < 1)
		{
			System.out.println("usage: java validation_custom logging mw-id=<middleware ID>");
			System.exit(-1);
		}

		Config config = new Config(args);

		initializeLogging(config);

		//o Enable Message validation via connection configuration
		config.addValue("gmsec-msg-content-validate-send", "true");
		config.addValue("gmsec-validation-level", "3");

		Log.info("API version: " + ConnectionManager.getAPIVersion());

		try
		{
			ConnectionManager connMgr = new ConnectionManager(config);

			connMgr.initialize();

			Log.info("Middleware version: " + connMgr.getLibraryVersion());

			//o Register custom message validator
			CustomMessageValidator cmv = new CustomMessageValidator();
			connMgr.registerMessageValidator(cmv);

			//o Set up standard/common fields used with all messages
			int specVersion = connMgr.getSpecification().getVersion();
			setupStandardFields(specVersion);

			//o Create Heartbeat Message
			//o Note: Message subject and schema ID vary depending on the specification in use
			final String subject  = (specVersion > gmsecMIST.GMSEC_ISD_2018_00 ? HB_MSG_SUBJECT : ALT_HB_MSG_SUBJECT);
			final String schemaID = (specVersion > gmsecMIST.GMSEC_ISD_2018_00 ? "MSG.HB" : "MSG.C2CX.HB");

			MistMessage msg = new MistMessage(subject, schemaID, connMgr.getSpecification());

			//o Add PUB-RATE field with illegal value
			msg.setValue("PUB-RATE", "5");

			//o For very old specifications, we need to add a MSG-ID field
			if (specVersion <= gmsecMIST.GMSEC_ISD_2014_00)
			{
				msg.addField("MSG-ID", "12345");
			}

			//o Attempt to publish malformed message
			try
			{
				Log.info("Attempting to publish malformed message...");
				connMgr.publish(msg);

				Log.warning("Was expecting an error");
			}
			catch (GMSEC_Exception e)
			{
				//o We expect to encounter error with the PUB-RATE field
				Log.info("This is an expected error:\n" + e.getMessage());
			}

			//o Fix PUB-RATE field with legal value
			msg.setValue("PUB-RATE", "15");

			//o Publish a good message
			try
			{
				Log.info("Attempting to publish good message...");
				connMgr.publish(msg);

				Log.info("Message published!");
			}
			catch (GMSEC_Exception e)
			{
				Log.warning("Unexpected error:\n" + e.getMessage());
			}

			//o Disconnect from the middleware and clean up the Connection
			connMgr.cleanup();

			//o Clear standard/common fields used with all messages
			clearStandardFields();
		}
		catch (GMSEC_Exception e)
		{
			Log.error(e.getMessage());
			System.exit(-1);
		}

		return;
	}

	public static void initializeLogging(Config config)
	{
		// If it was not specified in the command-line arguments, set
		// LOGLEVEL to 'INFO' and LOGFILE to 'stdout' to allow the
		// program report output on the terminal/command line
		String logLevel = config.getValue("LOGLEVEL");
		String logFile  = config.getValue("LOGFILE");

		if (logLevel == null)
		{
			config.addValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.addValue("LOGFILE", "STDOUT");
		}
	}

	private static void setupStandardFields(int specVersion)
		throws GMSEC_Exception
	{
		ArrayList<Field> definedFields = new ArrayList<Field>();

		StringField mission       = new StringField("MISSION-ID", "MISSION");
		StringField constellation = new StringField("CONSTELLATION-ID", "CONSTELLATION");
		StringField satIdPhys     = new StringField("SAT-ID-PHYSICAL", "SATELLITE");
		StringField satIdLog      = new StringField("SAT-ID-LOGICAL", "SATELLITE");
		StringField facility      = new StringField("FACILITY", "GMSEC-LAB");
		StringField component     = new StringField("COMPONENT", "VALIDATION-CUSTOM");
		StringField domain1       = new StringField("DOMAIN1", "DOMAIN1");
		StringField domain2       = new StringField("DOMAIN2", "DOMAIN2");

		definedFields.add(mission);
		definedFields.add(constellation);
		definedFields.add(satIdPhys);
		definedFields.add(satIdLog);
		definedFields.add(facility);
		definedFields.add(component);

		if (specVersion >= gmsecMIST.GMSEC_ISD_2018_00)
		{
			definedFields.add(domain1);
			definedFields.add(domain2);
		}

		MistMessage.setStandardFields(definedFields);
	}

	private static void clearStandardFields()
	{
		MistMessage.clearStandardFields();
	}
}

File: examples/csharp-dotnet/validation_custom/validation_custom.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/**
 * @file validation_custom.cs
 *
 * This file contains an example demonstrating how to implement additional
 * Message validation logic in addition to that which the GMSEC API provides.
 *
 * Note: This example focuses on adding additional validation upon the receipt
 * of a message.  It almost goes without saying that additional logic can be
 * added to a program prior to invoking the publish() function without having
 * to do anything special.
 */

using GMSEC.API;
using GMSEC.API.UTIL;
using GMSEC.API.MIST;
using GMSEC.API.MIST.MESSAGE;
using System;
using System.Collections.Generic;

class ValidationCallback: ConnectionManagerCallback
{
	public override void OnMessage(ConnectionManager connMgr, Message message)
	{
		try
		{
			//o Run the message through the GMSEC API-supplied validation
			connMgr.GetSpecification().ValidateMessage(message);

			//o In this example scenario, we are expecting to receive a
			// GMSEC PROD message containing a URI to a location on the disk
			// where a product file has been placed for retrieval.  In this
			// case, we want to validate that the location on the disk is
			// in an area which we would expect (i.e. Something that the
			// team has agreed upon prior to operational deployment).
			//
			// By validating this URI, we ensure that no malicious users
			// have infiltrated the system and somehow modified the messages
			// to cause us to retrieve a file from an unknown location.

			//o Start by checking to ensure that this is a PROD message
			if (IsProdMsg(message))
			{
				GMSEC.API.MIST.MESSAGE.ProductFileMessage prodMessage = new GMSEC.API.MIST.MESSAGE.ProductFileMessage(message.ToXML());

				//o Extract the Product File URI location(s) from the
				// message using a ProductFileIterator
				ProductFileIterator prodIter = prodMessage.GetProductFileIterator();

				while(prodIter.HasNext())
				{
					ProductFile prodFile = prodIter.Next();

					//o Check to ensure that the URI contains "//hostname/dir"
					String prodUri = prodFile.GetURI();
					if(prodUri.IndexOf("//hostname/dir") == -1)
					{
						String errorMsg = "Received an invalid PROD Message (bad URI):\n" + message.ToXML();
						throw new GMSEC_Exception(StatusClass.MIST_ERROR, StatusCode.MESSAGE_FAILED_VALIDATION, errorMsg);
					}
				}

				Log.Info("Received a valid message:\n" + message.ToXML());
			}
		}
		catch (Exception e)
		{
			Log.Error(e.ToString());
		}
	}

	public bool IsProdMsg(Message message)
	{
		bool result =
			message.GetStringValue("MESSAGE-TYPE") == "MSG"
			&& message.GetStringValue("MESSAGE-SUBTYPE") == "PROD";

		return result;
	}
}

class validation_custom
{
	static private String PROD_MESSAGE_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.PROD.PRODUCT_MESSAGE";

	static int Main(string[] args)
	{
		if (args.Length < 1)
		{
			Console.WriteLine("usage validation_custom.exe mw-id=<middleware ID>");
			return -1;
		}

		Config config = new Config(args);

		InitializeLogging(config);

		//o Enable Message validation.  This parameter is "false" by default.
		config.AddValue("GMSEC-MSG-CONTENT-VALIDATE", "true");

		Log.Info(ConnectionManager.GetAPIVersion());

		try
		{
			ConnectionManager connMgr = new ConnectionManager(config);

			Log.Info("Opening the connection to the middleware server");
			connMgr.Initialize();

			Log.Info(connMgr.GetLibraryVersion());

			//o Set up the ValidationCallback and subscribe
			ValidationCallback vc = new ValidationCallback();
			connMgr.Subscribe(PROD_MESSAGE_SUBJECT, vc);

			//o Start the AutoDispatcher
			connMgr.StartAutoDispatch();

			//o Create and publish a simple Product File Message
			SetupStandardFields(connMgr);

			GMSEC.API.MIST.MESSAGE.ProductFileMessage productMessage = CreateProductFileMessage(connMgr, "//hostname/dir/filename");

			connMgr.Publish(productMessage);

			productMessage = CreateProductFileMessage(connMgr, "//badhost/dir/filename");

			//o Publish the message to the middleware bus
			connMgr.Publish(productMessage);

			connMgr.StopAutoDispatch();

			//o Disconnect from the middleware and clean up the Connection
			connMgr.Cleanup();
		}
		catch (GMSEC_Exception e)
		{
			Log.Error(e.ToString());
			return -1;
		}

		return 0;
	}

	static void InitializeLogging(Config config)
	{
		String logLevel = config.GetValue("LOGLEVEL");
		String logFile = config.GetValue("LOGFILE");

		if (logLevel == null)
		{
			config.AddValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.AddValue("LOGFILE", "STDERR");
		}
	}

	static void SetupStandardFields(ConnectionManager connMgr)
	{
		List<Field> definedFields = new List<Field>();

		StringField missionField = new StringField("MISSION-ID", "MISSION");
		StringField satIdField = new StringField("SAT-ID-PHYSICAL", "SPACECRAFT");
		StringField facilityField = new StringField("FACILITY", "GMSEC Lab");
		StringField componentField = new StringField("COMPONENT", "device_message");

		definedFields.Add(missionField);
		definedFields.Add(satIdField);
		definedFields.Add(facilityField);
		definedFields.Add(componentField);

		if (connMgr.GetSpecification().GetVersion() >= GMSEC_ISD.GMSEC_ISD_2018_00)
		{
			StringField domain1 = new StringField("DOMAIN1", "MY-DOMAIN-1");
			StringField domain2 = new StringField("DOMAIN2", "MY-DOMAIN-2");

			definedFields.Add(domain1);
			definedFields.Add(domain2);
		}

		connMgr.SetStandardFields(definedFields);
	}

	static GMSEC.API.MIST.MESSAGE.ProductFileMessage CreateProductFileMessage(ConnectionManager connMgr, String filePath)
	{
		ProductFile externalFile = new ProductFile("External File", "External File Description", "1.0.0", "TXT", filePath);

		GMSEC.API.MIST.MESSAGE.ProductFileMessage productMessage;

		if (connMgr.GetSpecification().GetVersion() <= GMSEC_ISD.GMSEC_ISD_2016_00)
		{
			productMessage = new GMSEC.API.MIST.MESSAGE.ProductFileMessage(PROD_MESSAGE_SUBJECT, ResponseStatus.SUCCESSFUL_COMPLETION, "MSG.PROD.AUTO", connMgr.GetSpecification());
		}
		else
		{
			productMessage = new GMSEC.API.MIST.MESSAGE.ProductFileMessage(PROD_MESSAGE_SUBJECT, ResponseStatus.SUCCESSFUL_COMPLETION, "MSG.PROD", connMgr.GetSpecification());

			productMessage.AddField(new StringField("PROD-TYPE", "AUTO"));
			productMessage.AddField(new StringField("PROD-SUBTYPE", "DM"));
		}

		productMessage.AddProductFile(externalFile);

		connMgr.AddStandardFields(productMessage);

		return productMessage;
	}
}
